File: 859. Buddy Strings/solution.js
/**
 * @param {string} s
 * @param {string} goal
 * @return {boolean}
 */
 var buddyStrings = function(s, goal) {
  if (s.length != goal.length){
    return false;    
  }
  let sSet = new Set(s);
  if (s == goal && sSet.size < s.length){
    return true;
  }

  let difference = [];
  for (let index = 0; index < s.length; index++){
    if (s[index] != goal[index]){
      difference.push(index);
    }
  }

  return difference.length == 2 && s[difference[0]] == goal[difference[1]] && s[difference[1]] == goal[difference[0]];
};

File: 859. Buddy Strings/solution.java
class Solution {
  public boolean buddyStrings(String s, String goal) {
    if (s.length() != goal.length()){
      return false;    
    }
    
    Set<Character> sSet = new HashSet<>();
    for (int index = 0; index < s.length(); index++){
      sSet.add(s.charAt(index));
    }
    
    if (s.equals(goal) && sSet.size() < s.length()){
      return true;
    }

    List<Integer> difference = new ArrayList<Integer>();
    for (int index = 0; index < s.length(); index++){
      if (s.charAt(index) != goal.charAt(index)){
        difference.add(index);
      }
    }

    return difference.size() == 2 && s.charAt(difference.get(0)) == goal.charAt(difference.get(1)) && s.charAt(difference.get(1)) == goal.charAt(difference.get(0));
  }
}